marlin-succinct/protocol/src/prover.rs

```diff
@@ -172,189 +172,171 @@ impl<E: PairingEngine> ProverProof<E>
- 
-                // evaluate H1, G1 polynomials
-                h1_eval = h1.evaluate(oracles.betta[0]);
-                g1_eval = g1.evaluate(oracles.betta[0]);
-
-                // query batch[0], betta[1] random oracles from the argument context
-                argument.commit_scalar(&h1_eval);
-                oracles.batch[0] = argument.challenge();
-                argument.commit_scalar(&g1_eval);
-                oracles.betta[1] = argument.challenge();
-                        
-                // commit to W, ZA, ZB, H1 & G1 polynomials and
-                // open the commitments in batch at the first random oracle
-                match
-                (
-                    index.urs.commit(&w, index.h_group.size()),
-                    index.urs.commit(&zv[0], index.h_group.size()),
-                    index.urs.commit(&zv[1], index.h_group.size()),
-                    index.urs.commit(&h1, index.h_group.size()*2),
-                    index.urs.commit(&g1, index.h_group.size()-1),
-                    index.urs.open_batch
-                    (
-                        &vec!
-                        [
-                            za,
-                            zb,
-                            w,
-                            h1,
-                            g1,
-                        ],
-                        oracles.batch[0],
-                        oracles.betta[0]
-                    )
-                )
-                {
-                    (Some(w), Some(za), Some(zb), Some(h1), Some(g1), Some(open)) =>
-                    {
-                        w_comm = w;
-                        za_comm = za;
-                        zb_comm = zb;
-                        h1_comm = h1;
-                        g1_comm = g1;
-                        proof1 = open;
-                    }
-                    (_,_,_,_,_,_) => return Err(ProofError::ProofCreation)
-                }
+                w_comm = w;
+                za_comm = za;
+                zb_comm = zb;
+                h1_comm = h1;
+                g1_comm = g1;
+                proof1 = open;
             }
-            _ => return Err(ProofError::PolyDivision)
+            (_,_,_,_,_,_) => return Err(ProofError::ProofCreation)
         }
 
         // compute commitments, evaluations and openings for the second sumcheck argument
         // proceed with the random oracle argument further to query additional oracles
-        // betta[2], batch[1]
+        // beta[2], batch[1]
+
+        let (h2, mut g2) = Self::sumcheck_2_compute (index, &ra, &oracles)?;
+        sigma2 = g2.coeffs[0];
+        g2.coeffs.remove(0);
+
+        // evaluate H2, G2 polynomials
+        h2_eval = h2.evaluate(oracles.beta[1]);
+        g2_eval = g2.evaluate(oracles.beta[1]);
+
+        // query batch[1], beta[2] random oracles from the argument context
+        argument.commit_scalar(&h2_eval);
+        oracles.batch[1] = argument.challenge();
```

Comment:
There is no need to have separate challenges for all the batches. You can just sample one at the end of the prover sending all the polynomial commitments and messages (I call it [xi here](https://github.com/CodaProtocol/coda/blob/double-marlin-verifier/src/lib/double_marlin/double_marlin.ml#L239))